Orange/data/table/base.py

```diff
@@ -0,0 +1,1242 @@
+        meta : bool, default False
+            Whether to include the domain metas in the result.
+        writable : bool, default False
+            Whether to mark the resulting domain as writable.
+        Returns
+        -------
+        np.ndarray
+            The numpy array of the selected and transformed table data.
+        """
+        cols = []
+        if X:
+            cols += self.domain.attributes
+        if Y:
+            cols += self.domain.class_vars
+        if meta:
+            cols += self.domain.metas
+
+        # support using this method in TableSeries, whose len gives the number of columns
+        # (because it's a row slice)
+        n_rows = 1 if isinstance(self, SeriesBase) else len(self)
+
+        # preallocate result, we fill it in-place
+        # we need a more general dtype for metas (commonly strings),
+        # otherwise assignment fails later
+        result = np.zeros((n_rows, len(cols)), dtype=object if meta else None)
+        # effectively a double for loop, see if this is a bottleneck later
+        for i, col in enumerate(cols):
+            if isinstance(self, SeriesBase):
+                # if this is used in TableSeries, we don't have a series but an element,
+                # because we are iterating over a row
+                result[:, i] = col.to_val(self[col])
+            else:
+                result[:, i] = col.to_val(self[col]).values
+        result.setflags(write=writable)
+        return result
+
+    @property
+    def X(self):
+        """Return a read-only 2D numpy array of X.
+
+        The columns are in the same order as the columns in Table.domain.attributes.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._to_numpy(X=True)
+
+    @property
+    def Y(self):
+        """Return a read-only numpy array of Y.
+
+        If there is only one column, a one-dimensional array is returned. Otherwise 2D.
+        The columns are in the same order as the columns in Table.domain.class_vars.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        result = self._to_numpy(Y=True)
+        return result[:, 0] if result.shape[1] == 1 else result
+
+    @property
+    def metas(self):
+        """Return a read-only 2D numpy array of metas.
+
+        The columns are in the same order as the columns in Table.domain.metas.
+
+        Returns
+        -------
+        np.ndarray
+        """
+        return self._to_numpy(meta=True)
+
+    @property
+    def weights(self):
+        """Get the weights as a 1D numpy array."""
+        if isinstance(self, SeriesBase):
+            # at least 1D when using this in TableSeries, which instead returns a 0D ndarray directly
+            result = np.atleast_1d(self[self._WEIGHTS_COLUMN])
+        else:
+            result = self[self._WEIGHTS_COLUMN].values
+        # even if weights are set with an integer, we need to return a float
```

Comment:
This is still open. Please [see below](https://github.com/biolab/orange3/pull/1347/files/#r74590968).